### PR TITLE
feat(optimizer): support apply join transpose and refactor correlated_input_ref

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Related Issue: #3842 #3862 #3860 #3976

- This PR handle more general case of subquery including correlated column in project, join condition and deeper subquery.
- Also I refactor the correlated_input_ref using `correlated_id` to absolutely positioning its outside relation instead of relatively positioning by `depth`.
- I use a new way to apply our Apply transpose rule in TopDown order util a fix-point When 2 or more Apply exist in a same query nested.
- I introduce ApplyJoinRule to handle apply join transpose in generally way which can further optimized later but not now for understandability. Also this rule can use PlanCorrelatedIdFinder to decide to bail out early to avoid duplicate the Apply operator.
- I restrict our PullUpCorrelatedPredicateRule by PlanCorrelatedIdFinder which can ensure there is no correlated_input_ref in deeper level.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* SQL commands, functions, and operators

### Release note

Support unnesting more deeper nested subquery and execution plan might changed for those query.

## Refer to a related PR or issue link (optional)

Related Issue: #3842 #3862 #3860 #3976

### CODE REVIEW

**Comment by hengm3467** [comment]
@chenzl25 Please provide a release note for the user-facing changes. Thanks!

**Comment by chenzl25** [comment]
> @chenzl25 Please provide a release note for the user-facing changes. Thanks!

Got it.

**Comment by xiangjinwu** [comment]
It feels like each item in the PR description can be in its own PR.

File: src/frontend/src/optimizer/rule/apply_scan.rs [hunk 1]
```diff
@@ -26,7 +26,8 @@ pub struct ApplyScanRule {}
 impl Rule for ApplyScanRule {
     fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
         let apply = plan.as_logical_apply()?;
-        let (left, right, on, join_type, correlated_indices) = apply.clone().decompose();
+        let (left, right, on, join_type, _correlated_id, correlated_indices) =
+            apply.clone().decompose();
         let apply_left_len = left.schema().len();
         assert_eq!(join_type, JoinType::Inner);
         // TODO: Push `LogicalApply` down `LogicalJoin`.
```
Comment (likg227) [inline src/frontend/src/optimizer/rule/apply_scan.rs:33]: please remove this TODO.

**Review by likg227 — COMMENTED** [review]
generally LGTM. 
BTW, how about adding some e2e tests?

**Comment by chenzl25** [comment]
> It feels like each item in the PR description can be in its own PR.

Should we split this PR into smaller ones? At that time the code is still under control.

**Comment by chenzl25** [comment]
> generally LGTM. BTW, how about adding some e2e tests?

Sure. e2e tests are on the way.

**Comment by xiangjinwu** [comment]
> Should we split this PR into smaller ones? At that time the code is still under control.

That would be good. So that it is easier to stay focused.

**Comment by chenzl25** [comment]
This PR is split into 4 PR. Now all of them have been merged.
- #4128 
- #4149 
- #4171  
- #4202

**Pull request closed without merging**